File: src/main/java/org/luxons/sevenwonders/game/boards/TradingRules.java
package org.luxons.sevenwonders.game.boards;

import java.util.EnumMap;
import java.util.Map;

import org.luxons.sevenwonders.game.resources.ResourceType;

public class TradingRules {

    private final Map<ResourceType, Map<Provider, Integer>> costs = new EnumMap<>(ResourceType.class);

    private final int defaultCost;

    public TradingRules(int defaultCost) {
        this.defaultCost = defaultCost;
    }

    public int getCost(ResourceType type, Provider provider) {
        return costs.computeIfAbsent(type, t -> new EnumMap<>(Provider.class)).getOrDefault(provider, defaultCost);
    }

    public void setCost(ResourceType type, Provider provider, int cost) {
        costs.computeIfAbsent(type, t -> new EnumMap<>(Provider.class)).put(provider, cost);
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/resources/TradingRules.kt
package org.luxons.sevenwonders.engine.resources

import org.luxons.sevenwonders.model.resources.Provider
import org.luxons.sevenwonders.model.resources.ResourceTransaction
import org.luxons.sevenwonders.model.resources.ResourceTransactions
import org.luxons.sevenwonders.model.resources.ResourceType

class TradingRules internal constructor(private val defaultCost: Int) {

    private val costs: MutableMap<ResourceType, MutableMap<Provider, Int>> = mutableMapOf()

    fun getCosts(): Map<ResourceType, Map<Provider, Int>> {
        return costs
    }

    internal fun getCost(type: ResourceType, provider: Provider): Int =
        costs.computeIfAbsent(type) { mutableMapOf() }.getOrDefault(provider, defaultCost)

    internal fun setCost(type: ResourceType, provider: Provider, cost: Int) {
        costs.computeIfAbsent(type) { mutableMapOf() }[provider] = cost
    }

    internal fun computeCost(transactions: ResourceTransactions): Int = transactions.map { computeCost(it) }.sum()

    internal fun computeCost(transact: ResourceTransaction) = computeCost(transact.asResources(), transact.provider)

    private fun computeCost(resources: Resources, provider: Provider): Int =
        resources.quantities.map { (type, qty) -> getCost(type, provider) * qty }.sum()
}
